app/src/main/java/com/kickstarter/viewmodels/SearchViewModel.java

```diff
@@ -23,94 +23,117 @@
+      super(environment);
+
+      final ApiClientType apiClient = environment.apiClient();
+      final Scheduler scheduler = environment.scheduler();
+
+      final Observable<DiscoveryParams> searchParams = search
+        .filter(StringUtils::isPresent)
+        .debounce(300, TimeUnit.MILLISECONDS, scheduler)
+        .map(s -> DiscoveryParams.builder().term(s).build());
+
+      final Observable<DiscoveryParams> popularParams = search
+        .filter(StringUtils::isEmpty)
+        .map(__ -> defaultParams)
+        .startWith(defaultParams);
+
+      final Observable<DiscoveryParams> params = Observable.merge(searchParams, popularParams);
+
+      final ApiPaginator<Project, DiscoverEnvelope, DiscoveryParams> paginator =
+        ApiPaginator.<Project, DiscoverEnvelope, DiscoveryParams>builder()
+          .nextPage(nextPage)
+          .startOverWith(params)
+          .envelopeToListOfData(DiscoverEnvelope::projects)
+          .envelopeToMoreUrl(env -> env.urls().api().moreProjects())
+          .clearWhenStartingOver(true)
+          .concater(ListUtils::concatDistinct)
+          .loadWithParams(apiClient::fetchProjects)
+          .loadWithPaginationPath(apiClient::fetchProjects)
+          .build();
+
+      search
+        .filter(StringUtils::isEmpty)
+        .compose(bindToLifecycle())
+        .subscribe(__ -> searchProjects.onNext(ListUtils.empty()));
+
+      params
+        .compose(Transformers.takePairWhen(paginator.paginatedData()))
+        .compose(bindToLifecycle())
+        .subscribe(paramsAndProjects -> {
+          if (paramsAndProjects.first.sort() == defaultSort) {
+            popularProjects.onNext(paramsAndProjects.second);
+          } else {
+            searchProjects.onNext(paramsAndProjects.second);
+          }
+        });
+
+      // Track us viewing this page
+      koala.trackSearchView();
+
+      // Track search results and pagination
+      final Observable<Integer> pageCount = paginator.loadingPage();
+      final Observable<String> query = params
+        .map(DiscoveryParams::term);
+      query
+        .compose(Transformers.takePairWhen(pageCount))
+        .filter(qp -> StringUtils.isPresent(qp.first))
+        .compose(bindToLifecycle())
+        .subscribe(qp -> koala.trackSearchResults(qp.first, qp.second));
+    }
+
+    private final PublishSubject<String> search = PublishSubject.create();
+    private final PublishSubject<Void> nextPage = PublishSubject.create();
+    private final PublishSubject<Project> projectCardClicked = PublishSubject.create();
+
+    private final BehaviorSubject<List<Project>> popularProjects = BehaviorSubject.create();
+    private final BehaviorSubject<List<Project>> searchProjects = BehaviorSubject.create();
+
+    public final SearchViewModel.Inputs inputs = this;
+    public final SearchViewModel.Outputs outputs = this;
+
+    @Override public void search(final @NonNull String s) {
+      search.onNext(s);
+    }
+
```

Comment:
we like to remove new lines between input and output functions, [an example here of our ProfileViewModel](https://github.com/kickstarter/android-oss/blob/master/app/src/main/java/com/kickstarter/viewmodels/ProfileViewModel.java#L175)